lib/Target/ExportSystemC/Patterns/EmitCEmissionPatterns.cpp

```diff
@@ -26,11 +27,181 @@ namespace {
+    p << applyOp.getApplicableOperator();
+    p.getInlinable(applyOp.getOperand())
+        .emitWithParensOnLowerPrecedence(Precedence::ADDRESS_OF);
+  }
+};
+
+/// Emit emitc.call operations. Only calls with 0 or 1 results are supported.
+/// Calls with no result are emitted as statements whereas calls with exactly
+/// one result are always inlined no matter whether it is a pure function or has
+/// side effects. To make sure that calls with side effects are not reordered
+/// with interferring operations, a pre-pass has to emit VariableOp operations
+/// with the result of the call as initial value.
+class CallOpEmitter : public OpEmissionPattern<CallOp> {
+public:
+  using OpEmissionPattern::OpEmissionPattern;
+
+  MatchResult matchInlinable(Value value) override {
+    if (auto callOp = value.getDefiningOp<CallOp>()) {
+      // TODO: template arguments not supported for now.
+      if (callOp->getNumResults() == 1 && !callOp.getTemplateArgs())
+        return Precedence::FUNCTION_CALL;
+    }
+    return {};
+  }
+
+  void emitInlined(Value value, EmissionPrinter &p) override {
+    printCallOp(value.getDefiningOp<CallOp>(), p);
+  }
+
+  bool matchStatement(Operation *op) override {
+    // TODO: template arguments not supported for now.
+    if (auto callOp = dyn_cast<CallOp>(op))
+      return callOp->getNumResults() <= 1 && !callOp.getTemplateArgs();
+    return false;
+  }
+
+  void emitStatement(CallOp callOp, EmissionPrinter &p) override {
+    if (callOp->getNumResults() != 0)
+      return;
+
+    printCallOp(callOp, p);
+    p << ";\n";
+  }
+
+private:
+  void printCallOp(CallOp callOp, EmissionPrinter &p) {
+    p << callOp.getCallee();
+
+    p << "(";
+    bool first = true;
+
+    if (!callOp.getArgs()) {
+      for (Value operand : callOp.getOperands()) {
+        if (!first)
+          p << ", ";
+
+        p.getInlinable(operand).emitWithParensOnLowerPrecedence(
+            Precedence::COMMA);
+        first = false;
+      }
+    } else {
+      for (Attribute attr : callOp.getArgs().value()) {
+        if (!first)
+          p << ", ";
+
+        first = false;
+
+        if (auto idx = attr.dyn_cast<IntegerAttr>()) {
+          if (idx.getType().isa<IndexType>()) {
+            p.getInlinable(callOp.getOperands()[idx.getInt()])
+                .emitWithParensOnLowerPrecedence(Precedence::COMMA);
+            continue;
+          }
+        }
+
+        p.emitAttr(attr);
+      }
+    }
+
+    p << ")";
+  }
+};
+
+/// Emit emitc.cast operations.
+struct CastOpEmitter : OpEmissionPattern<CastOp> {
+  using OpEmissionPattern::OpEmissionPattern;
+
+  MatchResult matchInlinable(Value value) override {
+    if (value.getDefiningOp<CastOp>())
+      return Precedence::CAST;
+    return {};
+  }
+
+  void emitInlined(Value value, EmissionPrinter &p) override {
+    auto castOp = value.getDefiningOp<CastOp>();
+    p << "(";
+    p.emitType(castOp.getDest().getType());
+    p << ") ";
```

Comment:
This is a C-style cast. This seems fine for `emitc::CastOp`. Are there situations where this should be emitted as a C++-style `static_cast`?

Reply:
Indeed, in many cases it's preferable to emit a `static_cast` to have the improved safety of additional compiler checks, but a C-style cast seemed 'more default' to me. Ideally, there would be an attribute on `emitc.cast` that specifies which of the several casts that C++ supports should be used.